src/Compilers/Core/Portable/SourceGeneration/Nodes/NodeStateTable.cs

```diff
@@ -163,37 +162,63 @@ public NodeStateTable<T> CreateCachedTableWithUpdatedSteps<TInput>(NodeStateTabl
 
         public sealed class Builder
         {
-            private readonly ArrayBuilder<TableEntry> _states;
+            /// <summary>
+            /// Only mutate this field from within <see cref="AddState(TableEntry)"/>.
+            /// </summary>
+            private readonly ArrayBuilder<TableEntry> _states_doNotMutateDirectly;
             private readonly NodeStateTable<T> _previous;
 
+            private readonly IEqualityComparer<T> _equalityComparer;
+
             private readonly string? _name;
             private readonly ArrayBuilder<IncrementalGeneratorRunStep>? _steps;
 
             [MemberNotNullWhen(true, nameof(_steps))]
             public bool TrackIncrementalSteps => _steps is not null;
 
-            internal Builder(NodeStateTable<T> previous, string? name, bool stepTrackingEnabled)
+            /// <summary>
+            /// Tracks if the new table we're building has all the same entries as the previous table.  If we end up
+            /// with the same set of entries at the end, we can just point our new table at that same array, avoiding a
+            /// costly allocation.
+            /// </summary>
+            private bool _unchangedFromPrevious = true;
+
+            internal Builder(NodeStateTable<T> previous, string? name, bool stepTrackingEnabled, IEqualityComparer<T>? equalityComparer)
             {
-                _states = ArrayBuilder<TableEntry>.GetInstance();
+                _states_doNotMutateDirectly = ArrayBuilder<TableEntry>.GetInstance(previous.Count);
                 _previous = previous;
                 _name = name;
+                _equalityComparer = equalityComparer ?? EqualityComparer<T>.Default;
                 if (stepTrackingEnabled)
                 {
                     _steps = ArrayBuilder<IncrementalGeneratorRunStep>.GetInstance();
                 }
             }
 
+            private void AddState(TableEntry entry)
+            {
+                var currentindex = _states_doNotMutateDirectly.Count;
+                _states_doNotMutateDirectly.Add(entry);
```

Comment:
❔ Can we avoid initializing `_states_doNotMutateDirectly` until adding the first item which is not the same as before? Overall this is trivia, but you'll need special handling for the case where the only change is the last item is removed.

Reply:
that's definitely possible, but likely more complex.  could def look into it in a future PR.